staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go

```diff
@@ -225,36 +303,78 @@ func Start(nodeServer interface{}, opts ...Option) (result DRAPlugin, finalErr e
 		return nil, fmt.Errorf("start node client: %v", err)
 	}
 	d.plugin = plugin
-	defer func() {
-		// Clean up if we didn't finish succcessfully.
-		if r := recover(); r != nil {
-			plugin.stop()
-			panic(r)
-		}
-		if finalErr != nil {
-			plugin.stop()
-		}
-	}()
 	if !implemented {
 		return nil, errors.New("no supported DRA gRPC API is implemented and enabled")
 	}
 
 	// Now make it available to kubelet.
-	registrar, err := startRegistrar(klog.LoggerWithName(o.logger, "registrar"), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, o.draAddress, o.pluginRegistrationEndpoint)
+	registrar, err := startRegistrar(klog.NewContext(ctx, klog.LoggerWithName(logger, "registrar")), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, o.draAddress, o.pluginRegistrationEndpoint)
 	if err != nil {
 		return nil, fmt.Errorf("start registrar: %v", err)
 	}
 	d.registrar = registrar
 
+	// startGRPCServer and startRegistrar don't implement cancellation
+	// themselves, we add that for both here.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		<-ctx.Done()
+		d.plugin.stop()
+		d.registrar.stop()
+	}()
+
 	return d, nil
 }
 
+func (d *draPlugin) PublishResources(ctx context.Context, nodeResources []*resourceapi.ResourceModel) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	d.publishResources = true
+	d.nodeResources = nodeResources
+	d.updatePublishResources()
+}
+
+// updatePublishResources is called each time something changed that might
+// affect how resources are published.
+func (d *draPlugin) updatePublishResources() {
```

Comment:
Can you explain the prupose of this function a bit more (meaning why can't its contents just be pulled directly into `PublishResources()`. That seems to be the only place it is being called, and it feel strange taht one might call this multiple times and end up starting multiple resourceSliceControllers.

Reply:
It does a delayed start of the controller. If `PublishResources` doesn't get called, the controller doesn't need to be started because it's unknown what it should sync.

> end up starting multiple resourceSliceControllers.

Ooops! That wasn't the intent. It should have checked whether there is one already and if so, update it. Will fix.

Reply:
I think this code used to be a bit different. There's no reason anymore for having this separate `updatePublishResources`, I'll remove it.

Reply:
I think I had options for initializing the plugin with some resources without calling `PublishResources`, then removed them because there's no need to have two APIs for doing the same thing.

I hope the revised code is clearer: https://github.com/kubernetes/kubernetes/compare/896aada559999cbef105f4fecef5e7c3dc2e8d70..bee827b1b42e53895b73ee6febb30ddf611797a9